### ISSUE TITLE
More info about canceling rpc call options

### ISSUE THREAD

**Issue by KSDaemon** [wamp-proto/wamp-proto#140]
Right now in advanced profile there is only one string, about posible options:
CANCEL.Options.mode|string == "skip" | "kill" | "killnowait"

We should extend docs with explanation, what do this options mean, and how they should be done on dealer/callee side.

**Comment by gammazero** [comment]
Here is what I propose:

Cancellation behaves differently depending on the mode:

- `skip`: The pending call is canceled and ERROR is sent immediately back to the caller.  No INTERRUPT is sent to the callee and the result is discarded when received.
- `kill`: INTERRUPT is sent to the client, but ERROR is not returned to the caller until after the callee has responded to the canceled call.  In this case the caller may receive RESULT or ERROR depending whether the callee finishes processing the invocation or the interrupt first.
- `killnowait`: The pending call is canceled and ERROR is send immediately back to the caller.  INTERRUPT is sent to the callee and any response to the invocation or interrupt from the callee is discarded when received.

If the callee does not support call canceling, then behavior is `skip`.

**Comment by ecorm** [comment]
Should it be permitted to send `CANCEL` without specifying the mode? If so, which mode is assumed by default?

**Comment by meejah** [comment]
I think it makes sense to have a default.

I think `kill` makes the most sense as a default, as it's the "most complete" and will give you end-to-end cancelling (i.e. when the ERROR comes back you know the other side has actually-cancelled their call vs. completing it before they got the INTERRUPT).

**Comment by ecorm** [comment]
Should the `mode` option be propagated to the `INTERRUPT` message? The spec does not say. Is there a need for the callee to know under which `mode` an `INTERRUPT` is received?

**Comment by oberstet** [comment]
@meejah having a default of `kill`: +1. least surprising ("complete") for client

@ecorm IMO: nope. for the callee, an `INTERRUPT` message being received already holds the semantics by its very reception. the callee will only get the `INTERRUPT` for `kill` and `killnowait` - and in both cases any result that might be already in processing or even in-flight back to the router is discarded anyways - so why/how should the callee change its behavior?

**Comment by gammazero** [comment]
+1 for default of `kill`

**Comment by gammazero** [comment]
@oberstet Communicating the mode to the callee is useful:  When the mode is `killnowait`, the callee is informed that the dealer is not waiting for an error response.  This allows the callee to omit sending a response to the caller, that would only be discarded.  When the mode is `kill` the callee knows that the dealer is waiting for a response and one must be sent.

**Comment by gammazero** [comment]
I am also reconsidering my vote for `kill` as the default mode.  Usually calls are canceled because the callee is taking too long to respond.  The `kill` mode waits for the callee to respond that the call was canceled, which is probably not what is wanted when the problem is that the callee is not responding in a reasonable time (or not at all).

Generally, `killnowait` is what is wanted unless there is reason to think the callee will respond to the CANCEL quickly even though no response has been received for the CALL.

**Comment by oberstet** [comment]
> When the mode is killnowait, the callee is informed that the dealer is not waiting for an error response.

no, that's not currently the case: the dealer will expect an ERROR message from the callee in both cases!

**Comment by gammazero** [comment]
@oberstet If the ERROR message is necessary in both cases, then it seems the callee would behave identically regardless of the mode.

In the case of `killnowait`, is it essential that the dealer expect a response?  Seems like the dealer could clean up the pending call as soon as the CANCEL (mode=killnowait) was received, and then drop and RESULT or ERROR for an INVOCATION that does not exist.

**Comment by oberstet** [comment]
> In the case of killnowait, is it essential that the dealer expect a response?

On the one hand, yes, it is not "essential".

On the other hand: do we want to "break" the regular message sequencing for a call just for that specific case?

For example, receiving a RESULT/ERROR (at the dealer, from a callee) for a non-pending invocation is a protocol error, whereas NOT having a callee respond with ERROR in `killnowait`  case deviates from that.

Also, it would mean that the dealer gets no feedback that the INTERRUPT was actually received and processed at all.

IMO, we should stay away from this kind of "optimization" (NOT having the callee response with ERROR/RESULT under all call scenarios) ..

But yeah, these are the corner cases and behavioral details we definitely need to discuss .. I am reopening this issue and add it to the meta issue that collects the issues we need to nail to complete this feature (I created the "meta issue" to make sure we dont miss anything that was already raised)

**Comment by gammazero** [comment]
@oberstet Thank you for the detailed explanation.  I completely agree with the logic, about avoiding these kinds of optimizations.  I have a small change to make in a certain dealer implementation.

**Comment by gammazero** [comment]
@meejah Is there a mechanism to clean up calls that are never answered, other than waiting for the callee to leave the realm?  Consider that a call may be canceled because callee is not responding.

One way for the dealer to handle cleanup for mode of `killnowait` and `skip` is to relax the protocol error determination and have a rule that states: `when a dealer receives an invocation error or result, for a call that does not exist, the dealer assumes this is a response to a canceled call and drops/logs the message`.

Similarly, `if a client receives an INTERRUPT for an INVOCATION that does not exist, the client assumes that a RESULT was already returned for the INVOCATION, and drops/logs the INTERRUPT without responding to it`.

If those two rules are explicit exceptions to protocol error, then that relieves the dealer and callee from having to keep track of calls and invocations that have already been responded to.

**Comment by ecorm** [comment]
One thing to consider regarding the default cancel mode is that `skip` does not require the callee to support the call cancelling advanced feature. Perhaps the default mode should be the one that's most likely to be supported by all callee clients?

**Comment by gammazero** [comment]
My recommendation is for the default mode to be `skipnowait`, because CANCEL will likely happen because of already waiting too long for a router response.  Also, if the _callee_ does not support canceling, then the router will automatically treat this as `skip`.
That last part is already per spec (https://wamp-proto.org/static/rfc/draft-oberstet-hybi-crossbar-wamp.html#rfc.section.14.3.4.1) `If the callee does not support call canceling, then behavior is skip.`

I agree with @ecorm to leave it explicit if we cannot come to consensus.

**Comment by meejah** [comment]
@gammazero I'm not sure if I follow the reasoning for `skipnowait` to be the default? It's likely that the caller is giving up on the callee if we're trying to explicitly cancel, right? If callee doesn't support cancelling, then it'll become `skip` anyway (no matter what the default is).

The problem will be that if the callee is so sad that it "never" responds to the cancel from the router, the router might waste some space on state -- I'd say this is maybe "implementation detail"? (e.g. the router could have a timer and give up eventually?) Or should we make that explicit in the spec -- that is, if a cancel-supporting callee doesn't respond in X seconds, "something" happens (protocol error and kill the connection?). Practically speaking, if a callee supports canceling but doesn't respond in a reasonable amount of time, disconnecting it is probably the right solution.

All that said, I do like the simplicity for routers of making exceptions on the protocol errors. This does make me vaguely uneasy, tough. "Debugging might be hard" is one reason I can think of, but the "log it" would fix that. Maybe I can come up with other reasons :)
